src/lxc/cgroups/cgroup2_devices.c

```diff
@@ -596,9 +619,23 @@ bool bpf_cgroup_devices_attach(struct cgroup_ops *ops, struct lxc_list *devices)
 	return log_trace(true, "Attached bpf program");
 }
 
+/*
+ * TODO: Clarify semantics.
+ * Specifically, when a user switches the type of device program, i.e. switches
+ * from blocking all devices by default to allowing all devices by default or
+ * vica versa do we reactivate the devices we have recorded so far or not?
+ * Specific example: The user configures a device program that blocks all
+ * devices by default apart from a small list of devices such as /dev/zero and
+ * /dev/null. Now the user switches to a device program that allows all devices
+ * by default. Naturally we skip all specific devices since they are
+ * encompassed in the global allow rule. But now assume the user switches back
+ * to a device program that blocks all devices by default. Do we reactivate the
+ * previously specific allowed devices, i.e. do we grant access to /dev/zero
+ * and /dev/null? My gut feeling is no, but I'm not sure.
```

Comment:
From the runc point of view, I think it makes the most sense to make cgroupv1 and cgroupv2 act as similarly as possible, so you should start with a clean slate if you switch the type of the device program.

Reply:
@cyphar, please see
https://github.com/lxc/lxc/pull/3684/commits/9d444b79282288e8c1e40189aaaa81315900dca2